### PR TITLE
Improve `anti_unify_with_constraints`

### PR DESCRIPTION
**Pull request by nwatson22**
Needed for https://github.com/runtimeverification/evm-semantics/pull/1934.
Adds an additional option to `anti_unify_with_constraints` which adds an additional constraint which is an `_orBool_` of the two anti-unifying substitutions.
Adds a test for the method with this option which tests a new variable is created to abstract and the `orBool` constraint is added as described.

### CODE REVIEW

File: src/pyk/kast/inner.py [hunk 1]
```diff
@@ -136,6 +136,21 @@ def ml_pred(self) -> KInner:
...
         return ml_term
 
+    @property
+    def pred(self) -> KInner:
+        items = []
+        for k in self:
+            val = self[k]
+            if type(val) is not KVariable or val.name != KVariable(k).name:
+                items.append(KApply('_==K_', [KVariable(k), val]))
+
```
Comment (nwatson22) [inline src/pyk/kast/inner.py:146]:
```suggestion
```

File: src/pyk/kast/inner.py [hunk 2]
```diff
@@ -136,6 +136,20 @@ def ml_pred(self) -> KInner:
...
+        for k in self:
+            val = self[k]
+            if type(val) is not KVariable or val.name != KVariable(k).name:
+                items.append(KApply('_==K_', [KVariable(k), val]))
+        if len(items) == 0:
+            return KToken('true', 'Bool')
+        term = items[0]
+        for _i in items[1:]:
+            term = KApply('_andBool_', [term, _i])
+        return term
```
Comment (tothtamas28) [inline src/pyk/kast/inner.py:151]: Consider using `functools.reduce`:

```suggestion
        conjuncts = [
            KApply('_==K_', KVariable(name), val) for name, val in self.items()
            if type(val) is not KVariable or val.name != k
        ]
        if not items:
            return KToken('true', 'Bool')
        return reduce(KLabel('__andbool__'), conjuncts)
```

(Not tested.)

File: src/pyk/kast/manip.py [hunk 3]
```diff
@@ -598,8 +598,18 @@ def _rewrites_to_abstractions(_kast: KInner) -> KInner:
...
+    constrained_term_2: KInner,
+    implications: bool = False,
+    constraint_disjunct: bool = False,
+    abstracted_disjunct: bool = False,
 ) -> KInner:
+    def disjunction_from_substs(substs: list[Subst]) -> KInner:
+        term = substs[0].pred
+        for _i in substs[1:]:
+            term = KApply('_orBool_', [term, _i.pred])
+        return mlEqualsTrue(term)
```
Comment (tothtamas28) [inline src/pyk/kast/manip.py:611]: It looks like the special case of `len(substs) == 2` is sufficient here:

```suggestion
    def disjunction_from_substs(subst1: Subst, subst2: Subst) -> KInner:
        return mlEqualsTrue(KApply('_orBool_', (subst1.pred, subst2.pred)))
```

**Review by tothtamas28 — APPROVED** [review]

File: src/pyk/kast/manip.py (hunk 3 above)
Comment (tothtamas28) [reply src/pyk/kast/manip.py:611]: Also, if you want as compact a constraint as possible, you need to handle the special case of a `subst` being `true`.

File: src/pyk/kast/inner.py (hunk 2 above)
Comment (nwatson22) [reply src/pyk/kast/inner.py]: Now using `reduce`.

File: src/pyk/kast/manip.py (hunk 3 above)
Comment (nwatson22) [reply src/pyk/kast/manip.py]: Simplified as suggested, and also handles the case where `subst.pred` is true.

**Pull request merged**